diff -ur --new-file vice-2.4/autom4te.cache/output.0 vice-2.4+jamexit/autom4te.cache/output.0
diff -ur --new-file vice-2.4/src/6510core.c vice-2.4+jamexit/src/6510core.c
--- vice-2.4/src/6510core.c	2012-10-14 22:57:24.000000000 +0200
+++ vice-2.4+jamexit/src/6510core.c	2014-06-30 20:40:58.763316905 +0200
@@ -38,6 +38,7 @@
 #define CPU_STR "Main CPU"
 #endif
 
+#include "jam.h"
 #include "traps.h"
 
 #ifndef C64DTV
diff -ur --new-file vice-2.4/src/jam.c vice-2.4+jamexit/src/jam.c
--- vice-2.4/src/jam.c	1970-01-01 01:00:00.000000000 +0100
+++ vice-2.4+jamexit/src/jam.c	2014-06-30 20:43:40.633869673 +0200
@@ -0,0 +1,34 @@
+/*
+ * jam.c - JAM
+ *
+ * Written by
+ *  Andre Fachat <afachat@gmx.de>
+ *
+ * This file is part of VICE, the Versatile Commodore Emulator.
+ * See README for copyright notice.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ *  02111-1307  USA.
+ *
+ */
+
+#include <stdlib.h>
+
+void jam_exit(void) {
+
+	exit(-2);
+}
+
+
diff -ur --new-file vice-2.4/src/jam.h vice-2.4+jamexit/src/jam.h
--- vice-2.4/src/jam.h	1970-01-01 01:00:00.000000000 +0100
+++ vice-2.4+jamexit/src/jam.h	2014-06-30 20:41:57.906788122 +0200
@@ -0,0 +1,33 @@
+/*
+ * jam.h - definition for JAM
+ *
+ * Written by
+ *  Andre Fachat <afachat@gmx.de>
+ *
+ * This file is part of VICE, the Versatile Commodore Emulator.
+ * See README for copyright notice.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ *  02111-1307  USA.
+ *
+ */
+
+#ifndef JAM_H
+#define JAM_H
+
+void jam_exit(void);
+
+#endif
+
diff -ur --new-file vice-2.4/src/machine.c vice-2.4+jamexit/src/machine.c
--- vice-2.4/src/machine.c	2012-10-18 23:54:21.000000000 +0200
+++ vice-2.4+jamexit/src/machine.c	2014-06-30 22:05:07.942173825 +0200
@@ -84,6 +84,8 @@
     va_list ap;
     ui_jam_action_t ret;
 
+    return JAM_EXIT;
+#if 0
     if (ignore_jam > 0)
         return JAM_NONE;
 
@@ -114,6 +116,7 @@
 
     ignore_jam = 1;
     return JAM_NONE;
+#endif
 }
 
 static void machine_trigger_reset_internal(const unsigned int mode)
diff -ur --new-file vice-2.4/src/machine.h vice-2.4+jamexit/src/machine.h
--- vice-2.4/src/machine.h	2011-10-25 18:30:00.000000000 +0200
+++ vice-2.4+jamexit/src/machine.h	2014-06-30 20:37:20.689266635 +0200
@@ -157,6 +157,7 @@
 #define JAM_RESET      1
 #define JAM_HARD_RESET 2
 #define JAM_MONITOR    3
+#define JAM_EXIT       4
 unsigned int machine_jam(const char *format, ...);
 
 /* Update memory pointers if memory mapping has changed. */
diff -ur --new-file vice-2.4/src/maincpu.c vice-2.4+jamexit/src/maincpu.c
--- vice-2.4/src/maincpu.c	2012-10-20 19:44:16.000000000 +0200
+++ vice-2.4+jamexit/src/maincpu.c	2014-06-30 20:42:17.758610633 +0200
@@ -501,6 +501,9 @@
             monitor_startup(e_comp_space);                            \
             IMPORT_REGISTERS();                                       \
             break;                                                    \
+	  case JAM_EXIT:					      \
+	    jam_exit();						      \
+	    break;						      \
           default:                                                    \
             CLK++;                                                    \
         }                                                             \
diff -ur --new-file vice-2.4/src/Makefile.am vice-2.4+jamexit/src/Makefile.am
--- vice-2.4/src/Makefile.am	2012-10-18 19:15:10.000000000 +0200
+++ vice-2.4+jamexit/src/Makefile.am	2014-06-30 20:45:34.384852662 +0200
@@ -154,6 +154,7 @@
 	initcmdline.h \
 	interrupt.h \
 	ioutil.h \
+	jam.h \
 	joystick.h \
 	kbdbuf.h \
 	keyboard.h \
@@ -264,6 +265,7 @@
 	initcmdline.c \
 	interrupt.c \
 	ioutil.c \
+	jam.c \
 	joystick.c \
 	kbdbuf.c \
 	keyboard.c \
diff -ur --new-file vice-2.4/src/Makefile.in vice-2.4+jamexit/src/Makefile.in
--- vice-2.4/src/Makefile.in	2012-11-02 22:06:53.000000000 +0100
+++ vice-2.4+jamexit/src/Makefile.in	2014-06-30 21:53:56.680175361 +0200
@@ -99,7 +99,7 @@
 	embedded.$(OBJEXT) event.$(OBJEXT) findpath.$(OBJEXT) \
 	fliplist.$(OBJEXT) gcr.$(OBJEXT) info.$(OBJEXT) init.$(OBJEXT) \
 	initcmdline.$(OBJEXT) interrupt.$(OBJEXT) ioutil.$(OBJEXT) \
-	joystick.$(OBJEXT) kbdbuf.$(OBJEXT) keyboard.$(OBJEXT) \
+	jam.$(OBJEXT) joystick.$(OBJEXT) kbdbuf.$(OBJEXT) keyboard.$(OBJEXT) \
 	lib.$(OBJEXT) libm_math.$(OBJEXT) lightpen.$(OBJEXT) \
 	log.$(OBJEXT) machine-bus.$(OBJEXT) machine.$(OBJEXT) \
 	main.$(OBJEXT) network.$(OBJEXT) opencbmlib.$(OBJEXT) \
@@ -820,6 +820,7 @@
 	interrupt.c \
 	ioutil.c \
 	joystick.c \
+	jam.c \
 	kbdbuf.c \
 	keyboard.c \
 	lib.c \

